File: CodeArchieve/UOJ/Test-Round-1/83-水题出题人/problem_83/quick_sort.cpp
#include <stdio.h>

int n, counter;
int a[1000001];

void quick_sort_r(int l, int r){
	int i, j, x, y;
	counter++;
	i = l;
	counter++;
	j = r;
	counter++;
	x = a[(l + r) / 2];
	counter++;
	do{
		counter++;
		while(a[i] < x){
			counter++;
			i++;
		}
		while(x < a[j]){
			counter++;
			j--;
		}
		counter++;
		if(i <= j){
			counter++;
			y = a[i];
			counter++;
			a[i] = a[j];
			counter++;
			a[j] = y;
			counter++;
			i++;
			counter++;
			j--;
		}
	}while(i <= j);
	counter++;
	if(l < j){
		counter++;
		quick_sort_r(l, j);
	}
	counter++;
	if(i < r){
		counter++;
		quick_sort_r(i, r);
	}
}

void quick_sort_main(){
	int i;
	counter++;
	scanf("%d", &n);
	counter++;
	for(i = 1; i <= n; i++){
		counter++;
		scanf("%d", &a[i]);
	}
	counter++;
	quick_sort_r(1, n);
	counter++;
	for(i = 1; i <= n; i++){
		counter++;
		printf("%d ", a[i]);
	}
}

int main(){
	counter = 0;
	quick_sort_main();
	printf("\n");
	if(counter <= 2000000){
		printf("Accepted! The counter value is :%d\n", counter);
	}else{
		printf("Time Limit Exceeded because the counter value is :%d\n", counter);
	}
	return 0;
}

File: CodeArchieve/UOJ/Test-Round-1/83-水题出题人/problem_83/quick_sort.pas
program quick_sort;

var n, counter :longint;
var a :array[1..1000000] of longint;

procedure quick_sort_r(l, r :longint);
var i, j, x, y :longint;
begin
	inc(counter);
	i := l;
	inc(counter);
	j := r;
	inc(counter);
	x := a[(l + r) div 2];
	inc(counter);
	repeat
		inc(counter);
		while a[i] < x do begin
			inc(counter);
			inc(i);
		end;
		while x < a[j] do begin
			inc(counter);
			dec(j);
		end;
		inc(counter);
		if i <= j then begin
			inc(counter);
			y := a[i];
			inc(counter);
			a[i] := a[j];
			inc(counter);
			a[j] := y;
			inc(counter);
			inc(i);
			inc(counter);
			dec(j);
		end;
	until i > j;
	inc(counter);
	if l < j then begin
		inc(counter);
		quick_sort_r(l, j);
	end;
	inc(counter);
	if i < r then begin
		inc(counter);
		quick_sort_r(i, r);
	end;
end;

procedure quick_sort_main;
var i :longint;
begin
	inc(counter);
	readln(n);
	inc(counter);
	for i := 1 to n do begin
		inc(counter);
		readln(a[i]);
	end;
	inc(counter);
	quick_sort_r(1, n);
	inc(counter);
	for i := 1 to n do begin
		inc(counter);
		write(a[i], ' ');
	end;
end;

begin
	counter := 0;
	quick_sort_main;
	writeln;
	if counter <= 2000000 then begin
		writeln('Accepted! The counter value is :', counter);
	end else begin
		writeln('Time Limit Exceeded because the counter value is :', counter);
	end;
end.